tests/copylat.cpp

```diff
@@ -0,0 +1,273 @@
+              << "; Bus id: "
+              << std::hex 
+              << std::setfill('0') << std::setw(4) << dev_pci_domain_id
+              << ":" << std::setfill('0') << std::setw(2) << dev_pci_bus_id
+              << ":" << std::setfill('0') << std::setw(2) << dev_pci_device_id
+              << std::dec
+              << endl;
+    }
+    cout << "selecting device " << dev_id << endl;
+    ASSERTDRV(cuDeviceGet(&dev, dev_id));
+
+    CUcontext dev_ctx;
+    ASSERTDRV(cuDevicePrimaryCtxRetain(&dev_ctx, dev));
+    ASSERTDRV(cuCtxSetCurrent(dev_ctx));
+
+    CUdeviceptr d_A;
+    ASSERTDRV(cuMemAlloc(&d_A, size));
+    cout << "device ptr: 0x" << hex << d_A << dec << endl;
+    cout << "allocated size: " << size << endl;
+
+    unsigned int flag = 1;
+    ASSERTDRV(cuPointerSetAttribute(&flag, CU_POINTER_ATTRIBUTE_SYNC_MEMOPS, d_A));
+
+    uint32_t *init_buf = NULL;
+    uint32_t *h_buf = NULL;
+    ASSERTDRV(cuMemAllocHost((void **)&init_buf, size));
+    ASSERT_NEQ(init_buf, (void*)0);
+    ASSERTDRV(cuMemAllocHost((void **)&h_buf, size));
+    ASSERT_NEQ(h_buf, (void*)0);
+    init_hbuf_walking_bit(init_buf, size);
+
+    if (do_cumemcpy) {
+        cout << endl;
+        cout << "cuMemcpy_H2D num iters for each size: " << num_write_iters << endl;
+        printf("Test \t\t Size(B) \t Avg.Time(us)\n");
+        BEGIN_CHECK {
+            // cuMemcpy H2D benchmark
+            copy_size = 1;
+            while (copy_size <= size) {
+                int iter = 0;
+                clock_gettime(MYCLOCK, &beg);
+                for (iter = 0; iter < num_write_iters; ++iter) {
+                    ASSERTDRV(cuMemcpy(d_A, (CUdeviceptr)init_buf, copy_size));
+                }
+                clock_gettime(MYCLOCK, &end);
+                lat_us = ((end.tv_nsec-beg.tv_nsec)/1000.0 + (end.tv_sec-beg.tv_sec)*1000000.0) / (double)iter;
+                printf("cuMemcpy_H2D \t %8zu \t %11.4f\n", copy_size, lat_us);
+                copy_size <<= 1;
+            }
+        } END_CHECK;
+
+        cout << endl;
+        cout << "cuMemcpy_D2H num iters for each size: " << num_read_iters << endl;
+        printf("Test \t\t Size(B) \t Avg.Time(us)\n");
+        BEGIN_CHECK {
+            // cuMemcpy D2H benchmark
+            copy_size = 1;
+            while (copy_size <= size) {
+                int iter = 0;
+                clock_gettime(MYCLOCK, &beg);
+                for (iter = 0; iter < num_read_iters; ++iter) {
+                    ASSERTDRV(cuMemcpy((CUdeviceptr)h_buf, d_A, copy_size));
+                }
+                clock_gettime(MYCLOCK, &end);
+                lat_us = ((end.tv_nsec-beg.tv_nsec)/1000.0 + (end.tv_sec-beg.tv_sec)*1000000.0) / (double)iter;
+                printf("cuMemcpy_D2H \t %8zu \t %11.4f\n", copy_size, lat_us);
+                copy_size <<= 1;
+            }
+        } END_CHECK;
+
+        cout << endl;
+    }
+
+    cout << endl;
+
+    gdr_t g = gdr_open();
+    ASSERT_NEQ(g, (void*)0);
+
```

Comment:
please print a warning to the user re to_mapping not implying visibility to pre-launched GPU work, as stated in the GRDMA whitepaper